RRR4R: rational -> decimal

rigelrozanski opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺ 
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes. 
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  --> 
closes #1807

- [x] Linked to github-issue with discussion and accepted design
- [x] Updated all relevant documentation (`docs/`)
- [x] Updated all relevant code comments
- [x] Wrote tests
- [x] Added entries in `PENDING.md`
- [x] Updated `cmd/gaia` and `examples/`
___________________________________
For Admin Use:
- [x] Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- [x] Reviewers Assigned 
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

alexanderbez commented:
Thanks @rigelrozanski! Will review. Took a look at the build and it seems failures occur due to unexpected results probably related to the `Stringer` implementation of `Dec` vs the [old](https://golang.org/pkg/math/big/#Rat.String). I think we can just update the unit test require statements 👍

rigelrozanski commented:
ALRIGHT THIS IS READY 4 R E VI EW!

It's all passing they were just minor errors.
As mentioned previously there is one final task of justifying the decimal precision which I think should be addressed after this PR but before launch - this justification will be "hand" example calculations which go through core calculations - I think we should get the piggy bank fees in here before we do these example calculations as that one may be the limiting factor... for the time being 10 decimal points seems to be sufficient and should work on the testnets.   Opened an issue for this to follow up on prelaunch -> https://github.com/cosmos/cosmos-sdk/issues/1916

ValarDragon commented on client/lcd/lcd_test.go:
```diff
@@ -374,6 +374,7 @@ func TestValidatorsQuery(t *testing.T) {
 	require.True(t, foundVal, "pkBech %v, owner %v", pkBech, validators[0].Owner)
 }
 
+//XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXxx
```
Could a note be added here instead of a bunch of `X`'s?

rigelrozanski replied:
my bad, this was left over from debugging

ValarDragon commented on docs/resources/whitepaper.md:
```diff
@@ -821,7 +821,7 @@ in the case of a successful double-spend attack.
 Building on an approach pioneered by Ripple, Stellar [\[13\]][13] refined a
 model of Federated Byzantine Agreement wherein the processes participating in
-consensus do not constitute a fixed and globally known set. Rather, each
+consensus do not constitute a fixed and globally known set. Decher, each
```
sed gone wrong lol

rigelrozanski replied:
reverted all instances (and all cases)

ValarDragon commented on docs/sdk/core/app3.md:
```diff
@@ -312,7 +312,7 @@ accounts.
 kind of functionality each module gets access to. In particular, we try to
 follow the *principle of least authority*.
-Rather than providing full blown access to the `KVStore` or the `AccountMapper`, 
+Decher than providing full blown access to the `KVStore` or the `AccountMapper`, 
```
Sed caught something it shouldn't have XD

rigelrozanski replied:
addessed all

ValarDragon commented on docs/spec/ibc/appendices.md:
```diff
@@ -31,11 +31,11 @@ See [binary format as protobuf specification](./protobuf/queue.proto)
 With a known root hash (which is included in the block headers), the existence of a particular key/value in the tree can be proven by tracing the path to the value and revealing the (k-1) hashes for the paths not taken on each level ([[10](./references.md#10)]).
 
-There are a number of different implementations of this basic idea, using different hash functions, as well as prefixes to prevent second preimage attacks (differentiating leaf nodes from inner nodes). Rather than force all chains that wish to participate in IBC to use the same data store, we provide a data structure that can represent Merkle proofs from a variety of data stores, and provide for  ...
+There are a number of different implementations of this basic idea, using different hash functions, as well as prefixes to prevent second preimage attacks (differentiating leaf nodes from inner nodes). Decher than force all chains that wish to participate in IBC to use the same data store, we provide a data structure that can represent Merkle proofs from a variety of data stores, and provide for  ...
```
`Decher -> Rather`

rigelrozanski replied:
addessed all

ValarDragon commented on docs/spec/ibc/appendices.md:
```diff
@@ -31,11 +31,11 @@ See [binary format as protobuf specification](./protobuf/queue.proto)
 We generalize the left/right idiom to the concatenation a (possibly empty) fixed prefix, the (just calculated) last hash, and a (possibly empty) fixed suffix. We must only define two fields on each level and can represent any type, even a 16-ary Patricia tree, with this structure. One must only translate from the store's native proof to this format, and it can be verified by any chain, providing  ...
 
-The proof format also allows for chaining of trees, combining multiple Merkle stores into a "multi-store". Many applications (such as the EVM) define a data store with a large proof size for internal use. Rather than force them to change the store (impossible), or live with huge proofs (inefficient), we provide the possibility to express Merkle proofs connecting multiple subtrees. Thus, one could ...
+The proof format also allows for chaining of trees, combining multiple Merkle stores into a "multi-store". Many applications (such as the EVM) define a data store with a large proof size for internal use. Decher than force them to change the store (impossible), or live with huge proofs (inefficient), we provide the possibility to express Merkle proofs connecting multiple subtrees. Thus, one could ...
```
Decher strikes again!

rigelrozanski replied:
addessed all

ValarDragon commented on docs/spec/ibc/overview.md:
```diff
@@ -8,7 +8,7 @@ The IBC protocol creates a mechanism by which two replicated fault-tolerant stat
 The IBC protocol assumes that multiple applications are running on their own blockchain with their own state and own logic. Communication is achieved over an ordered message queue primitive, allowing the creation of complex inter-chain processes without trusted third parties.
 
-The message packets are not signed by one psuedonymous account, or even multiple, as in multi-signature sidechain implementations. Rather, IBC assigns authorization of the packets to the source blockchain's consensus algorithm, performing light-client style verification on the destination chain. The Byzantine-fault-tolerant properties of the underlying blockchains are preserved: a user transferri ...
+The message packets are not signed by one psuedonymous account, or even multiple, as in multi-signature sidechain implementations. Decher, IBC assigns authorization of the packets to the source blockchain's consensus algorithm, performing light-client style verification on the destination chain. The Byzantine-fault-tolerant properties of the underlying blockchains are preserved: a user transferri ...
```
Decher strikes again!

rigelrozanski replied:
addessed all

ValarDragon commented on docs/spec/provisioning/overview.md:
```diff
@@ -205,13 +205,13 @@ defined to be 13% per year, however the annual inflation is capped as between
+AnnualInflationDeceChange = (1 - bondedRatio / 0.67) * 0.13
 
-annualInflation += AnnualInflationRateChange
+annualInflation += AnnualInflationDeceChange
```
Can these parameters be changed back?

rigelrozanski replied:
Replaced all

ValarDragon commented on docs/spec/staking/end_block.md:
```diff
@@ -42,16 +42,16 @@ processProvisions():
    
-    inflationRateChangePerYear = (1 - bondedRatio / params.GoalBonded) * params.InflationRateChange
-    inflationRateChange = inflationRateChangePerYear / hrsPerYr
+    inflationDeceChangePerYear = (1 - bondedRatio / params.GoalBonded) * params.InflationDeceChange
```
More Find/Replace issues

rigelrozanski replied:
addressed

ValarDragon commented on docs/spec/staking/state.md:
```diff
@@ -29,10 +29,10 @@ overall functioning of the stake module.
-	InflationMax        sdk.Rat // maximum inflation rate
-	InflationMin        sdk.Rat // minimum inflation rate
-	GoalBonded          sdk.Rat // Goal of percent bonded atoms
+    InflationDeceChange sdk.Dec // maximum annual change in inflation rate
```
`Dece` strikes again!

rigelrozanski replied:
Looked through everything `Dece` and made fixes

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+// create a new Dec from integer assuming whole numbers
+// CONTRACT: prec !> Precision
```
should the contract be: `prec <= Precision` or `Precision > prec`, the `!>` looks odd to me?

rigelrozanski replied:
sure, updated to `prec <= Precision`

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+// get the precision multiplier
+func precisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
```
If we validate all created decimals, is there a reason for this check?

rigelrozanski replied:
it's a sanity check, hence a panic not an error return. Going to leave in here for now (we can remove or disable sanity for optimization later in the future if we want)

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+}
+
+// number of decimal places
+const Precision = 10
```
I assume these are base 10 decimal places. Have we benchmarked the performance gains we'd get with base 2 here? It may be worth considering. (Though I guess we could hard fork and change this #postlaunch)

rigelrozanski replied:
Check it out rat.Big doesn't store things in any bases - it only presents things in the desired base when you call getting a string... so I think it's already optimized! However I will quote that we determined during our call (Jae, you and myself I believe) that we cannot use base 2 due to some rounding problems for user interaction.  - but again I don't think this issue matters because optimized in `big.Int` -> https://golang.org/pkg/math/big/#Int.Text

ValarDragon replied:
Your totally right! Thanks for the clarification

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+func (d Dec) IsZero() bool      { return (d.Int).Sign() == 0 } // Is equal to zero
+func (d Dec) Equal(d2 Dec) bool { return (d.Int).Cmp(d2.Int) == 0 }
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }             // greater than
+func (d Dec) GTE(d2 Dec) bool   { return !d.LT(d2) }                            // greater than or equal
```
Can this be switched to 
`(d.Int).Cmp(d2.Int) > -1`?
I'd prefer to not use a function call here, for better efficiency as its reasonable to assume that decimals will be in tight loops.

rigelrozanski replied:
oh yeah this is a good idea.. thanks! included

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }             // greater than
+func (d Dec) GTE(d2 Dec) bool   { return !d.LT(d2) }                            // greater than or equal
+func (d Dec) LT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == -1 }            // less than
+func (d Dec) LTE(d2 Dec) bool   { return !d.GT(d2) }                            // less than or equal
```
Can this be switched to
(d.Int).Cmp(d2.Int) <1?
I'd prefer to not use a function call here, for better efficiency as its reasonable to assume that decimals will be in tight loops.

rigelrozanski replied:
thanks! good idear agian

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+const Precision = 10
+
+func precisionInt() *big.Int {
+	return new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
```
Lets cache this result in an `init` function, and just use `new(big.Int).Set` between calls. #postlaunch incorporate a big int pool for usage just within this file.

rigelrozanski replied:
added something here to reuse for precision - I think a lot of reuse potential in this file just need to do it right, different In operations assign things in weird ways - kind of shitty UX in big's side

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+
+// quotient
+func (d Dec) Quo(d2 Dec) Dec {
+	mul := new(big.Int).Mul(new(big.Int).Mul( // multiple Precision twice
```
We shouldn't create two seperate big ints here.
Instead this would be better as:
```
mul :=new(big.Int).Mul(d.Int, precisionInt())
mul.Mul(mul, d.precisionInt())
```

rigelrozanski replied:
added

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+
+	quo := new(big.Int).Quo(mul, d2.Int)
+	chopped := BankerRoundChop(quo, Precision)
+	return Dec{chopped}
```
Are we getting the right power of 10 in the result here? It doesn't seem like it would to me.

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,324 @@
+// multiplication
+func (d Dec) Mul(d2 Dec) Dec {
+	mul := new(big.Int).Mul(d.Int, d2.Int)
+	chopped := BankerRoundChop(mul, Precision)
```
Lets check for overflow here, like we do with normal Sdk Ints. (We don't want it to be over 256 bits + prec_bits long)

rigelrozanski replied:
How do we calculate the bits used for 10 decimals of precision? haven't resolved yet (feel free to push fyi over the weekend - but I'll get to it on tuesday otherwise)

ValarDragon replied:
If we want a fast heuristic, we can cache the base2 log of `precision`, and set `prec_bits = ceil(log_2(precision))`. (We could then do this heuristic before computation as well),

rigelrozanski replied:
cool - yeah the 33.2 makes sense to me, (appreciate the explanation!), I think the simplest thing is to perform the computation on the resulting value. doing this we can just use the value of `chopped` which already has 10 decimal places thus the check now becomes:

```golang
	if chopped.BitLen() > 289 {
		panic("Int overflow")
	}
```

This way we can use the same overflow check for all computations without having to consider how to determine the bytes required for each operation type based on the length of the input variables. (as in you're adding the bytes for a multiplication, what about for the rest of BEDMAS family?)

rigelrozanski replied:
addressed added in these sanity checks for Add/Mul/Quo/Sub as well as RoundInt64

ValarDragon commented on x/stake/types/validator.go:
```diff
@@ -182,21 +182,21 @@ type BechValidator struct {
-	CommissionChangeToday sdk.Rat `json:"commission_change_today"` // XXX commission rate change today, reset each day (UTC time)
+	Commission            sdk.Dec `json:"commission"`              // XXX the commission rate of fees charged to any delegators
+	CommissionMax         sdk.Dec `json:"commission_max"`          // XXX maximum commission rate which this validator can ever charge
+	CommissionChangeDece  sdk.Dec `json:"commission_change_rate"`  // XXX maximum daily increase of the validator commission
```
find/replace errors strike again!

rigelrozanski replied:
Addressed

zmanian commented on docs/spec/staking/state.md:
```diff
@@ -90,14 +90,14 @@ type Validator struct {
-    Max         sdk.Rat  // maximum commission rate which this validator can ever charge
-    ChangeRate  sdk.Rat  // maximum daily increase of the validator commission
-    ChangeToday sdk.Rat  // commission rate change today, reset each day (UTC time)
+    Dece        sdk.Dec  // the commission rate of fees charged to any delegators
```
Probably want a better variable name here

rigelrozanski replied:
was a find replace error - changed back to `Rate`

ValarDragon commented on client/lcd/test_helpers.go:
```diff
@@ -173,7 +173,7 @@ func InitializeTestLCD(t *testing.T, nValidators int, initAddrs []sdk.AccAddress
 		acc := gapp.NewGenesisAccount(&accAuth)
 		genesisState.Accounts = append(genesisState.Accounts, acc)
-		genesisState.StakeData.Pool.LooseTokens = genesisState.StakeData.Pool.LooseTokens.Add(sdk.NewRat(100))
+		genesisState.StakeData.Pool.LooseTokens = genesisState.StakeData.Pool.LooseTokens.Add(sdk.NewDec(100, 0))
```
This doesn't need to block this PR, but how about a `NewDecFromInt` function just for this.

ValarDragon commented:
Just made some commits to improve the efficiency of this.

`precisionInt` no longer recomputes the precisionInt every time. Instead it just copies the bytes from precisionReuse into a new big int. precisionMultipliers are also now cached.

We now use `precisionReuse` instead of creating a new big int, internally where it won't be mutated. This saves the (very non-neglible) memory allocation time.

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+}
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{big.NewInt(0)} }
```
Do we need this to be a function - why not a constant (since all the mutating functions copy anyways)? Would save some allocation.

ValarDragon replied:
An external caller could do `ZeroDec().Int.Add(one, one)`.

However this should cache the big.NewInt(0), and just call a copy.

rigelrozanski replied:
addressed

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+
+// nolint - common values
+func ZeroDec() Dec { return Dec{big.NewInt(0)} }
+func OneDec() Dec  { return Dec{precisionInt()} }
```
Ditto.

ValarDragon replied:
See my previous comment

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+}
+
+// create a new Dec from big integer assuming whole numbers
+// CONTRACT: prec !> Precision
```
What's "!>"? `prec <= Precision`, maybe?

rigelrozanski replied:
addressed

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+	if d.Sign() == -1 {
+		d = new(big.Int).Neg(d)
+		defer func() {
+			chopped = new(big.Int).Neg(chopped)
```
Can we do this in-place?

ValarDragon replied:
Also, can we avoid the defer? thats a 10-20 ns overhead. Non-neglible here.

rigelrozanski replied:
oh good to know about the overhead, didn't realize thanks. - (addressed by Val)

rigelrozanski replied:
addressed by val. val: clever iterative solution, thanks.

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+
+	// remove the negative and add it back when returning
+	if d.Sign() == -1 {
+		d = new(big.Int).Neg(d)
```
Can we do this in-place?

rigelrozanski replied:
I originally had the logic flow to consider negatives throughout it (instead of performing only for the positive and then making it a negative again at the end) but that logic was actually a big headache, it's much cleaner to mutate and then unmutate, and I think its easier to understand given all the additionally if statements required if we don't do this. Had to do it this way with rat too. -> going to leave

rigelrozanski replied:
addressed by val

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+		}
+
+		// always round to an even number
+		if finalDig == 0 || finalDig == 2 || finalDig == 4 ||
```
`finalDig % 2 == 0`?

(or byte comparision if switched re: above)

ValarDragon replied:
Byte comparison seems like the way to go. Then we could just do `finalDig & 1 == 0`

rigelrozanski replied:
cool thanks - addressed

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+		return
+	default: // bankers rounding must take place
+		str := quo.String()
+		finalDig, err := strconv.Atoi(string(str[len(str)-1]))
```
Why do we need to convert to a string and then to an integer - can't we just check the byte value directly?

rigelrozanski replied:
addressed by val

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+			chopped = quo
+			return
+		}
+		chopped = new(big.Int).Add(quo, big.NewInt(1))
```
`quo` can be reused here I think

rigelrozanski replied:
addressed

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+		chopped = quo
+		return
+	case 1:
+		chopped = new(big.Int).Add(quo, big.NewInt(1))
```
`quo` can be reused here I think

rigelrozanski replied:
addressed

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+		return quo
+	}
+
+	lenWhole := len(d.String())
```
Do we really need to convert to a string here? Shouldn't the length be calculable - `Precision` (characters after decimal), plus one for `.` (decimal), plus `floor (log_10 d)` (characters before the decimal), or something like that?

rigelrozanski replied:
did a bunch of research there isn't an obvious way of doing this... the only thing I could see was converting to a float64 which is far less than ideal. doesn't look like there are log functions for big.Int (or big.Float!) but either way I think this conversion is fine and efficient - remember this is simply a big.Int.String() not a dec.String() which means it doesn't have a decimal point or any of the fun logic in dec.String().

I'm going to leave this

ValarDragon replied:
Im confused why we need strings. Why is the result different from:
```
BankerStatus := rem.Cmp(5*10^9)
If BankerStatus == -1 {
 Return quo 
} if BankerStatus == 0 {
 Parity check 
} else 
{ 
Return quo + 1
}
```
No strings necessary, right? (Sorry for bad syntax, on mobile)

rigelrozanski replied:
yes!  ----- yeah okay so there was a subtle change that was made (and probs should have been communicated better, but more importantly caught by a linter) which was to modify this function to instead chopping off `n` amount of digits from the end - I actually fully agree with this change as a good point of optimization because we actually ONLY ever round by the precision. -> what the linter should have caught was the fact that `n` was still included as a function input even though it was not used anywhere -> if we do have a linter for this I suspect it wasn't caught because we DO have one recursive function call in here, so even though the variable isn't used for any calculations, it's actually used to pass on to the next recursion (sneaky!). But yeah this function probably should have been fully refactored with that in mind when it was updated to only ever chop a `Precision` number of digits ---------------- BUT you're totally right, it really don't need a string or any of the "fun" logic to determine the number of leading zeros in the remainder. Thanks! updating now. (👏  woot less complexity )

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,356 @@
+}
+
+// create a decimal from a decimal string (ex. "1234.5678")
+func NewDecFromStr(str string) (d Dec, err Error) {
```
This function needs to be fuzzed (since it sanitizes user input)

ValarDragon replied:
Fuzzing this isn't critical, since we validate all returned values makes sense.

rigelrozanski replied:
I second val

ValarDragon commented on types/decimal.go:
```diff
@@ -209,11 +211,12 @@ func (d Dec) ToLeftPadded(totalDigits int8) string {
 	// remove the negative and add it back when returning
-	if d.Sign() == -1 {
+	if negated {
 		d = new(big.Int).Neg(d)
```
This comment got hidden I think,
Chris asked if we can do this in place, and I asked if we could do this without the defer.

rigelrozanski replied:
clever iterative solution, thanks.

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,351 @@
+	precisionReuse       = new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+	precisionMultipliers []*big.Int
+	zeroInt              = big.NewInt(0)
+	tenInt               = big.NewInt(10)
```
Oh lol I didn't see ten

rigelrozanski replied:
deleted `ten` and other unused var from before val's changes. Thanks

ValarDragon replied:
Thanks, my bad re that var!

zmanian commented:
I realized that when this is merged I am going to have to massage the state dump `gaia-700x` from fractions to decimals for `gaia-8000`

alexanderbez commented:
Correct @zmanian

AdityaSripal commented on types/decimal.go:
```diff
@@ -0,0 +1,351 @@
+
+// create a new Dec from integer assuming whole numbers
+// CONTRACT: prec <= Precision
+func NewDec(i, prec int64) Dec {
```
Is there a reason why we're requiring precision this time, compared to the approach with rationals which was to assume denominator=1 if not specified. Seems like we can do the same here with default precision=0

ValarDragon replied:
I agree. We should make the default param prec is 0, and then a second function with newDecWithPrec

rigelrozanski replied:
meh - okay

rigelrozanski replied:
k added in

AdityaSripal commented on types/decimal_test.go:
```diff
@@ -0,0 +1,298 @@
+func TestNewDecFromStr(t *testing.T) {
+	largeBigInt, success := new(big.Int).SetString("3144605511029693144278234343371835", 10)
+	require.True(t, success)
+	tests := []struct {
```
Include test for negative numbers as well as for malformed negative numbers
e.g. "0.-45"

rigelrozanski replied:
negative testcases do exist, they are added in each time. But I'll add in a malformed negative test case

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,351 @@
+// calculate the precision multiplier
+func calcPrecisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
```
sure added

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,351 @@
+// get the precision multiplier. Do not mutate result.
+func precisionMultiplier(prec int64) *big.Int {
+	if prec > Precision {
+		panic("too much precision")
```
added

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,351 @@
+}
+
+// create a new Dec from integer assuming whole numbers
+// CONTRACT: prec <= Precision
```
👍

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,351 @@
+}
+
+// create a decimal from a decimal string (ex. "1234.5678")
+func NewDecFromStr(str string) (d Dec, err Error) {
```
added bunch more description here

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,351 @@
+//   BankerRoundChop(1005, 1) = 100
+//   BankerRoundChop(1015, 1) = 102
+//   BankerRoundChop(1500, 3) = 2
+func BankerRoundChop(d *big.Int, n int64) (chopped *big.Int) {
```
hmm it looks like val probably modified this function to just use the precision factor by default which I think is fine actually, so I'm going to remove `n` from here (darn tootin' linter should have caught that). Thanks for the catch.

Also fair point on the use of the word chop, I'm going to change to just be `BankersRound`

rigelrozanski commented on x/fee_distribution/keeper_test.go:
```diff
@@ -8,8 +8,8 @@ package stake
 //candidatesIn[i] = NewCandidate(addrVals[i], pks[i], Description{})
-//candidatesIn[i].BondedShares = sdk.NewRat(amt)
-//candidatesIn[i].DelegatorShares = sdk.NewRat(amt)
+//candidatesIn[i].BondedShares = sdk.NewDec(amt)
```
distribution has changed significantly enough that this early code is no longer relevant, deleting it all.

rigelrozanski commented:
@cwgoes @ValarDragon @alexanderbez @AdityaSripal 

I've addressed all your comments, still needs a big rebase before merge which I'll do tomorrow (also not compiling because of a function name change which I'll update at the same time)

Thanks for the reviews!

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,351 @@
+}
+
+// intended to be used with require/assert:  require.True(DecEq(...))
+func DecEq(t *testing.T, exp, got Dec) (*testing.T, bool, string, Dec, Dec) {
```
Lol whoops I forgot there was a reason why this was here - you can't reference functions you've defined in a `xxx_test.go` file outside of that package - it needs to in some kind of a utilities file if we want to separate it out (like `test_common.go`)... since this is only one small function I'm going to leave it where it was for now

ValarDragon commented on types/decimal.go:
```diff
@@ -25,6 +26,7 @@ const (
 
 var (
 	precisionReuse       = new(big.Int).Exp(big.NewInt(10), big.NewInt(Precision), nil)
+	fivePrecision        = new(big.Int).Mul(big.NewInt(5), precisionReuse)
```
Shouldn't this be `(5 * 10**9)`?

rigelrozanski replied:
fixed

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+func (d Dec) GTE(d2 Dec) bool   { return (d.Int).Cmp(d2.Int) >= 0 }     // greater than or equal
+func (d Dec) LT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == -1 }    // less than
+func (d Dec) LTE(d2 Dec) bool   { return (d.Int).Cmp(d2.Int) <= 0 }     // less than or equal
+func (d Dec) Neg() Dec          { return Dec{new(big.Int).Neg(d.Int)} } // Is equal to zero
```
The comment here is wrong

rigelrozanski replied:
thanks addressed

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+// quotient
+func (d Dec) Quo(d2 Dec) Dec {
+
+	// multiply precision twice
```
Do we need two multiplications here? We should be able to do one multiplication and then a pure bankers round. Don't need to block this PR on this though.

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+		chopped = quo
+		return
+	case 1:
+		chopped = quo.Add(quo, big.NewInt(1))
```
We should cache the one

cwgoes replied:
++

rigelrozanski replied:
cached (and the one below)

ValarDragon commented:
Also do we want to align / rename int /uint's `NewIntWithDecimal` functions? I'd prefer to actually just remove those methods, and have the caller do them.

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+
+	// first extract any negative symbol
+	neg := false
+	if string(str[0]) == "-" {
```
Can we just compare the character directly?

rigelrozanski replied:
addressed

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+
+	//zerosToAdd := int64(lenRem - 1 + leadingZeros)
+	//multiplier := new(big.Int).Exp(big.NewInt(10), big.NewInt(zerosToAdd), nil)
+	//fiveLine := new(big.Int).Mul(big.NewInt(5), multiplier)
```
Agreed, git has history, we don't need to keep comments

rigelrozanski replied:
yeah meant to delete - this should just be deleted never need it again

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+// wraps d.MarshalText()
+func (d Dec) MarshalAmino() (string, error) {
+	if d.Int == nil {
+		d.Int = new(big.Int)
```
Can we return a constant string here?

rigelrozanski replied:
confused - expand?

ValarDragon replied:
That is what he was suggesting AFAICT.

rigelrozanski replied:
k - addressed - just define the zero values once now above

cwgoes commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+// MarshalJSON defines custom encoding scheme
+func (d Dec) MarshalJSON() ([]byte, error) {
+	if d.Int == nil {
+		d.Int = new(big.Int)
```
Can we return a constant string here?

rigelrozanski replied:
addressed

melekes commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+	mul.Mul(mul, precisionReuse)
+
+	quo := new(big.Int).Quo(mul, d2.Int)
+	chopped := ChopPrecisionAndRound(quo)
```
what about this check?

```
if chopped.BitLen() > 255+DecimalPrecisionBytes {
		panic("Int overflow")
	}
```

rigelrozanski replied:
thanks - added

melekes commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+}
+
+// intended to be used with require/assert:  require.True(DecEq(...))
+func DecEq(t *testing.T, exp, got Dec) (*testing.T, bool, string, Dec, Dec) {
```
maybe move it to `_test.go` file?

rigelrozanski replied:
cannot, see: https://github.com/cosmos/cosmos-sdk/pull/1819#discussion_r209438898

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+//
+// NOTE an error will return if more decimal places
+// are provided in the string than the constant Precision
+func NewDecFromStr(str string) (d Dec, err Error) {
```
addressed - this does not modify the string (or copy it even)

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+	"testing"
+)
+
+// NOTE: never use new(Dec) or else we will panic unmarshalling into the
```
Concrete structs get a bit better efficiency than interfaces. (Also follows the stdlib's standard)

rigelrozanski replied:
k just leaving for now

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+// nolint - go-cyclo
+// Remove a Precision amount of rightmost digits and perform bankers rounding
+// on the remainder (gaussian rounding) on the digits which have been removed.
+func ChopPrecisionAndRound(d *big.Int) (chopped *big.Int) {
```
yeah it doesn't need to be exposed - updated this func to mutate `d` now

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+
+	combined, ok := new(big.Int).SetString(combinedStr, 10)
+	if !ok {
+		return d, ErrUnknownRequest("bad string to integer conversion")
```
k done

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,403 @@
+	}
+
+	if lenDecs > Precision {
+		return d, ErrUnknownRequest("too much Precision in decimal")
```
yup added that and the lenDec now

ValarDragon commented on types/decimal.go:
```diff
@@ -0,0 +1,422 @@
+
+	// get the trucated quotient and remainder
+	quo, rem := big.NewInt(0), big.NewInt(0)
+	quo, rem = quo.QuoRem(d, precisionReuse, rem)
```
We actually don't need to use `quo` here, it can just be `d`, and then `quo` never had to be allocated :)

rigelrozanski replied:
cool I'm going to make `quo, rem := d, big.NewInt(0)` just to keep it explicit - makes the return statements more clear too.

rigelrozanski replied:
okay - turns out this causes a few undesired side effects because often this function is expected to return a new decimal not modify the original - I'm going to leave as is for now, I feel like this is a pretty minor optimization - but feel free to make the modification and add the fixes (a bunch of lcd stuff fails)

ValarDragon replied:
Sure! I'll look into this more once we get get the main PR merged!

jaekwon commented on types/decimal.go:
```diff
@@ -0,0 +1,422 @@
+	return multiplier
+}
+
+// get the precision multiplier. Do not mutate result.
```
Maybe a way to make Dec immutable so we can enforce it?

ValarDragon replied:
These are only used within this file, so we don't need to add a separate method to enforce all decimals are immutable.

rigelrozanski replied:
leaving for now

jaekwon commented on types/decimal.go:
```diff
@@ -0,0 +1,422 @@
+//nolint
+func (d Dec) IsZero() bool      { return (d.Int).Sign() == 0 } // Is equal to zero
+func (d Dec) Equal(d2 Dec) bool { return (d.Int).Cmp(d2.Int) == 0 }
+func (d Dec) GT(d2 Dec) bool    { return (d.Int).Cmp(d2.Int) == 1 }     // greater than
```
I always use comparisons to zero, since the direction of comparison works on the values themselves.
(d.Int).Cmp(d2.Int) > 0 means d.Int > d2.Int, imo easier to grok visually.

rigelrozanski replied:
yup I like it changed

jaekwon commented on types/decimal.go:
```diff
@@ -0,0 +1,422 @@
+// |         |      __
+// |       | |   __|__|__
+// |_____:  /   | $$$    |
+//              |________|
```
lul.

rigelrozanski commented on types/decimal.go:
```diff
@@ -0,0 +1,422 @@
+
+	// bytes required to represent the above precision
+	// ceil(log2(9999999999))
+	DecimalPrecisionBytes = 34
```
thanks, addressed

rigelrozanski commented:
Merging - we can make further optimizations / adjustments in future PRs
